### PR TITLE
Allow ActOnArgsContainer to support protocols.act_on

### LINKED ISSUE #4368: Simulators preemptively entangle all qubits in any CircuitOperation
**Description of the issue**

When simulating a CircuitOperation, simulators preemptively entangle all the qubits there and leave them entangled afterwards. So even if a circuit operation does not have any entangling operations inside it, the simulators will leave the whole thing in an entangled state.

**How to fix the issue**

I think this can be fixed by adding a `CircuitOperation._act_on_` that runs the `protocols.act_on` for each op in the decomposition. Additionally `protocols.act_on` should accept an `OperationTarget` instead of just `ActOnArgs`, and `ActOnArgsContainer` should implement `_act_on_fallback_` that calls `apply_operation`. Finally, `_core_iterator` should call `protocols.act_on(op, sim_state)` instead of `sim_state.apply_op(op)`.

The resulting order of how the engine will attempt to operate will look like:
```
protocols.act_on(circuit_op, container) ->
circuit_op._act_on_(container) ->
# then for each op in circuit_op.decompose it'll try each of the following in order, and stops once one succeeds:
  op._act_on_(container)
  container._act_on_fallback_(op)
  # we only merge states if we get past here
  op._act_on_(container.args[qubits])
  container.args[qubits]._act_on_fallback_(op)
```

I am currently working on a fix for this. It's not a high-priority bug, but soundness here will make thinking about classical control easier, as we'll want entanglement smartness in our classical control ops as well.

### PR DESCRIPTION
**Pull request by daxfohl**
Fixes #4368

Previously ActOnArgsContainer was "the top" in the act_on stack. However this caused CircuitOperations to entangle all qubits they contain. Now ActOnArgsContainer is just another act_on target, and CircuitOperaiton.act_on pushes its sub-operations through the act_on protocol into the ActOnArgsContainer, so the container never entangles the whole subcircuit.

To note, this also represents something of a directional shift as we move toward classical flow control. Rather than the smarts living in the simulators, the smarts can live in the operations, such that the classical operations are given control of running their components. This approach isn't entirely new (stabilizer operations already use act_on to manage their own simulation within Clifford simulators), but specifically for classical control (of which subcircuits will be, themselves able to contain classical control operations), I anticipate this is the approach we will standardize on.

### CODE REVIEW

File: cirq-core/cirq/protocols/act_on_protocol.py [hunk 1]
```diff
@@ -147,7 +150,7 @@ def act_on(
 
     arg_fallback = getattr(args, '_act_on_fallback_', None)
     if arg_fallback is not None:
-        qubits = action.qubits if is_op else qubits
+        qubits = action.qubits if isinstance(action, ops.Operation) else qubits
```
Comment (MichaelBroughton) [inline cirq-core/cirq/protocols/act_on_protocol.py:153]: Why the change here ?

File: cirq-core/cirq/sim/act_on_args_container.py [hunk 2]
```diff
@@ -77,29 +77,35 @@ def create_merged_state(self) -> TActOnArgs:
...
-        op: 'cirq.Operation',
+        action: Union['cirq.Operation', 'cirq.Gate'],
+        qubits: Sequence['cirq.Qid'],
+        allow_decompose: bool = True,
     ):
-        gate = op.gate
+        if isinstance(action, ops.Operation):
+            gate = action.gate
+        else:
+            gate = action
```
Comment (MichaelBroughton) [inline cirq-core/cirq/sim/act_on_args_container.py:89]:
```python3
gate = action
if isinstance(action, ops.Operation):
   gate = action.gate

```

**Review by MichaelBroughton — COMMENTED** [review]
Looks good. Few small nits. Also think we might be missing some return type annotations on the `_act_on_fallback_`s in a lot of places. I would appreciate if @95-martin-orion took one final pass on this since he's a little more familiar than I am.

File: cirq-core/cirq/ops/measurement_gate.py [hunk 3]
```diff
@@ -238,7 +238,11 @@ def _from_json_dict_(cls, num_qubits, key, invert_mask, qid_shape=None, **kwargs
     def _has_stabilizer_effect_(self) -> Optional[bool]:
         return True
 
-    def _act_on_(self, args: 'cirq.ActOnArgs', qubits: Sequence['cirq.Qid']) -> bool:
+    def _act_on_(self, args: 'cirq.OperationTarget', qubits: Sequence['cirq.Qid']) -> bool:
+        from cirq.sim import ActOnArgs
+
+        if not isinstance(args, ActOnArgs):
+            return NotImplemented
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/ops/measurement_gate.py:245]: A lot of the protocol dunder methods work like this. Honestly I'm not a fan of the pattern and it feels hacky and type-unsafe to me, but when in Rome. (I'd love to work on pushing things more toward a structured, typesafe design, but given how things are now, it's almost an entirely new product at that point).

File: cirq-core/cirq/contrib/quimb/mps_simulator.py [hunk 4]
```diff
@@ -440,9 +440,16 @@ def apply_op(self, op: 'cirq.Operation', prng: np.random.RandomState):
...
+    def _act_on_fallback_(
+        self,
+        action: Union['cirq.Operation', 'cirq.Gate'],
+        qubits: Sequence['cirq.Qid'],
+        allow_decompose: bool = True,
+    ):
         """Delegates the action to self.apply_op"""
-        return self.apply_op(op, self.prng)
+        if isinstance(action, ops.Gate):
+            action = ops.GateOperation(action, qubits)
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/contrib/quimb/mps_simulator.py:451]: There's already an apply_op in the MPS simulator that has this signature, so this was the least invasive fix.

I tried a couple months ago to separate act_on into two distinct protocols, one for ops and one for gates, but that got met with some resistance. So now there's some flopping back and forth between gates and ops within implementations of the protocol. I'll spend a little time looking into whether there's a better way to consolidate. It may make sense to do this promotion right as the first step of protocols.act_on itself. But obviously outside the scope of this PR.

File: cirq-core/cirq/protocols/act_on_protocol.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/protocols/act_on_protocol.py:153]: mypy suddenly couldn't infer that `is_op` implies that `action.qubits` existed.

File: cirq-core/cirq/ops/measurement_gate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply cirq-core/cirq/ops/measurement_gate.py:245]: Protocol methods in cirq often look something like this (pseudocode below):
```
For an object k and methods _m1_.._mN_
    If k has method _m1_, return k._m1_() if it does not return NotImplemented
    Repeat for _m2_ to _mN_
    If nothing has been returned yet, return an error
```
This "return NotImplemented" strategy is consistent with many builtin methods, like `__eq__`.

File: cirq-core/cirq/sim/act_on_args_container.py [hunk 5]
```diff
@@ -77,29 +77,35 @@ def create_merged_state(self) -> TActOnArgs:
...
+
         if isinstance(gate, ops.IdentityGate):
-            return
+            return True
```
Comment (95-martin-orion) [inline cirq-core/cirq/sim/act_on_args_container.py:92]: If this method has a return type, it should be documented in the function header.

File: cirq-core/cirq/sim/act_on_args_container_test.py [hunk 6]
```diff
@@ -98,6 +103,15 @@ def test_entanglement_causes_join():
     assert args[None] is not args[q0]
 
 
+def test_subcircuit_entanglement_causes_join():
```
Comment (95-martin-orion) [inline cirq-core/cirq/sim/act_on_args_container_test.py:106]: It would be nice to have a couple of additional tests in this space, e.g.:
- CircuitOp can disentangle qubits
- CircuitOp can (dis-)entangle a subset of the qubits it acts on

**Comment by 95-martin-orion** [comment]
Only a couple of items to add on my part.

File: cirq-core/cirq/sim/act_on_args_container.py (hunk 5 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args_container.py:89]: Added to base docs, also added return type spec to all implementations

File: cirq-core/cirq/contrib/quimb/mps_simulator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/contrib/quimb/mps_simulator.py:451]: I opened it as a new issue in https://github.com/quantumlib/Cirq/issues/4400

**Comment by 95-martin-orion** [comment]
Heads up: I'll be unavailable starting next week, so let me know if you'd like a re-review on this before then.

**Comment by daxfohl** [comment]
It's ready, but not urgent.

**Review by MichaelBroughton — APPROVED** [review]
LGTM
